js/web/lib/wasm/jsep/backend-webnn.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// WebNN API currently does not have a TypeScript definition file. This file is a workaround with types generated from
+// WebNN API specification.
+// https://github.com/webmachinelearning/webnn/issues/677
+/// <reference path="webnn/webnn.d.ts" />
+
+import { Tensor } from 'onnxruntime-common';
+
+import { DataType } from '../wasm-common';
+import { getInstance } from '../wasm-factory';
+
+import { createView } from './tensor-view';
+import { BufferId, createBufferManager } from './webnn/buffer-manager';
+import { LOG_DEBUG } from './log';
+
+/*
+ * TensorProto::data_type to WebNN OperandType mapping.
+ */
+const onnxDataTypeToWebnnDataType = new Map<DataType, MLOperandDataType>([
+  [DataType.float, 'float32'],
+  [DataType.float16, 'float16'],
+  [DataType.int32, 'int32'],
+  [DataType.uint32, 'uint32'],
+  [DataType.int64, 'int64'],
+  [DataType.uint64, 'uint64'],
+  [DataType.int8, 'int8'],
+  [DataType.uint8, 'uint8'],
+  [DataType.bool, 'uint8'],
+]);
+
+/**
+ * WebNN backend implementation. This class is used to keep track of the MLBuffers created by the backend and keep track
+ * of the current MLContext being used by the sessions.
+ */
+export class WebNNBackend {
+  /**
+   * Buffer managers for each session.
+   */
+  private bufferManager = createBufferManager(this);
+  /**
+   * Maps from session id to MLContexts.
+   */
+  private mlContextBySessionId = new Map<number, MLContext>();
+  /**
+   * Maps from MLContext to session ids.
+   */
+  private sessionIdsByMLContext = new Map<MLContext, Set<number>>();
+  /**
+   * Current session id.
+   */
+  private activeSessionId?: number;
+
+  public get currentSessionId(): number {
+    if (this.activeSessionId === undefined) {
+      throw new Error('No active session');
+    }
+    return this.activeSessionId;
+  }
+
+  public onRunStart(sessionId: number): void {
+    this.activeSessionId = sessionId;
+  }
+
+  public get currentContext(): MLContext {
+    return this.getMLContext(this.currentSessionId);
+  }
+
+  public registerMLContext(sessionId: number, mlContext: MLContext): void {
+    this.mlContextBySessionId.set(sessionId, mlContext);
+    let sessionIds = this.sessionIdsByMLContext.get(mlContext);
+    if (!sessionIds) {
+      sessionIds = new Set();
+      this.sessionIdsByMLContext.set(mlContext, sessionIds);
+    }
+    sessionIds.add(sessionId);
+  }
+
+  public unregisterMLContext(sessionId: number): void {
+    const mlContext = this.mlContextBySessionId.get(sessionId)!;
+    if (!mlContext) {
+      throw new Error(`No MLContext found for session ${sessionId}`);
```

Comment:
>  Error: No MLContext found for session 538579648

I happen to get this error, when running WebNN EP first, then Wasm EP, during the call to `wasm.jsepOnReleaseSession`.

Wasm EP can also be used with JSEP, we should make sure it is WebNN EP to throw this error when release the session.

Reply:
The cleanup logic was kind of strange. The previous version seemed to have always passed `undefined` to the buffer manager.